File: Liftr/ExternalRouteSyncService.swift
import CoreLocation
import Foundation
import HealthKit
import Supabase

struct ExternalRoutePoint: Decodable, Sendable {
    let lat: Double
    let lon: Double
    let alt: Double?
    let ts: String?
}

struct ExternalRouteJob: Decodable, Sendable, Identifiable {
    let id: UUID
    let provider: String
    let provider_activity_id: String
    let started_at: Date
    let ended_at: Date?
    let duration_sec: Int?
    let activity_code: String?
    let route_points: [ExternalRoutePoint]
}

struct ExternalRouteSyncSummary: Sendable {
    var applied: Int = 0
    var skipped: Int = 0
    var failed: Int = 0
    var errorMessages: [String] = []

    static let unavailable = ExternalRouteSyncSummary(
        failed: 1,
        errorMessages: [HealthKitRouteWriteError.healthDataNotAvailable.localizedDescription]
    )
}

enum WearableRedirect {
    static let host = "wearable-callback"

    static func isWearableCallback(_ url: URL) -> Bool {
        url.host == host
    }

    static func status(from url: URL) -> String? {
        URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .queryItems?
            .first(where: { $0.name == "status" })?
            .value
    }

    static func provider(from url: URL) -> String? {
        URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .queryItems?
            .first(where: { $0.name == "provider" })?
            .value
    }
}

final class ExternalRouteSyncService {
    static let shared = ExternalRouteSyncService()

    private let syncEnabledKey = "externalRouteSyncEnabled"
    private let lastSyncAtKey = "externalRouteLastSyncAt"
    private let routesAppliedKey = "externalRouteAppliedCount"

    var isSyncEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: syncEnabledKey) }
        set { UserDefaults.standard.set(newValue, forKey: syncEnabledKey) }
    }

    var lastSyncAt: Date? {
        UserDefaults.standard.object(forKey: lastSyncAtKey) as? Date
    }

    var routesAppliedCount: Int {
        UserDefaults.standard.integer(forKey: routesAppliedKey)
    }

    @discardableResult
    func processPendingJobs(userId: UUID) async -> ExternalRouteSyncSummary {
        guard isSyncEnabled else { return ExternalRouteSyncSummary() }
        guard HealthKitRouteWriteService.shared.isHealthDataAvailable else {
            return .unavailable
        }

        var summary = ExternalRouteSyncSummary()
        do {
            try await HealthKitRouteWriteService.shared.requestRouteWriteAuthorization()
        } catch {
            summary.failed += 1
            summary.errorMessages.append(error.localizedDescription)
            return summary
        }

        let jobs: [ExternalRouteJob]
        do {
            jobs = try await fetchPendingJobs()
        } catch {
            summary.failed += 1
            summary.errorMessages.append(error.localizedDescription)
            return summary
        }

        for job in jobs {
            let partial = await processJob(job, userId: userId)
            summary.applied += partial.applied
            summary.skipped += partial.skipped
            summary.failed += partial.failed
            summary.errorMessages.append(contentsOf: partial.errorMessages)
        }

        if summary.applied > 0 {
            let total = routesAppliedCount + summary.applied
            UserDefaults.standard.set(total, forKey: routesAppliedKey)
        }
        UserDefaults.standard.set(Date(), forKey: lastSyncAtKey)
        return summary
    }

    func handleWearableCallbackURL(_ url: URL) async {
        guard WearableRedirect.isWearableCallback(url) else { return }
        guard WearableRedirect.status(from: url) == "connected" else { return }
        guard let userId = SupabaseManager.shared.client.auth.currentUser?.id else { return }
        isSyncEnabled = true
        _ = await processPendingJobs(userId: userId)
    }

    private func processJob(_ job: ExternalRouteJob, userId: UUID) async -> ExternalRouteSyncSummary {
        var summary = ExternalRouteSyncSummary()
        guard let activityCode = job.activity_code,
              let hkActivity = ExternalRouteActivityMapping.hkWorkoutActivityType(for: activityCode)
        else {
            await updateJobStatus(job.id, status: "failed", reason: "unsupported_activity", hkUUID: nil)
            summary.failed += 1
            return summary
        }

        let durationSec = job.duration_sec ?? max(1, Int(job.ended_at?.timeIntervalSince(job.started_at) ?? 60))
        let criteria = HealthKitWorkoutMatchCriteria(
            startedAt: job.started_at,
            endedAt: job.ended_at,
            durationSec: durationSec,
            activityType: hkActivity,
            preferredSourceBundleIds: ExternalRouteSourceBundles.preferredBundles(for: job.provider)
        )

        let windowStart = job.started_at.addingTimeInterval(-HealthKitWorkoutMatcher.startTolerance)
        let windowEnd = (job.ended_at ?? job.started_at).addingTimeInterval(HealthKitWorkoutMatcher.startTolerance)

        let workouts: [HKWorkout]
        do {
            workouts = try await HealthKitRouteWriteService.shared.fetchWorkouts(from: windowStart, to: windowEnd)
        } catch {
            await updateJobStatus(job.id, status: "failed", reason: "healthkit_query_failed", hkUUID: nil)
            summary.failed += 1
            summary.errorMessages.append(error.localizedDescription)
            return summary
        }

        var routeCounts: [UUID: Int] = [:]
        for workout in workouts {
            let routes = (try? await HealthKitCardioImportService.shared.workoutRoutes(for: workout)) ?? []
            routeCounts[workout.uuid] = routes.count
        }

        guard let workout = HealthKitWorkoutMatcher.findBestMatch(
            in: workouts,
            criteria: criteria,
            existingRouteCounts: routeCounts
        ) else {
            await updateJobStatus(job.id, status: "failed", reason: "no_matching_hk_workout", hkUUID: nil)
            summary.failed += 1
            return summary
        }

        let existingRoutes = (try? await HealthKitCardioImportService.shared.workoutRoutes(for: workout)) ?? []
        if ExternalRouteDedupPolicy.shouldSkip(
            workout: workout,
            existingRoutes: existingRoutes,
            provider: job.provider,
            providerActivityId: job.provider_activity_id
        ) {
            await updateJobStatus(job.id, status: "skipped", reason: "route_already_present", hkUUID: workout.uuid.uuidString.lowercased())
            summary.skipped += 1
            return summary
        }

        let locations = locations(from: job.route_points)
        guard locations.count >= 2 else {
            await updateJobStatus(job.id, status: "failed", reason: "insufficient_route_points", hkUUID: nil)
            summary.failed += 1
            return summary
        }

        let metadata = ExternalRouteDedupPolicy.routeMetadata(
            provider: job.provider,
            providerActivityId: job.provider_activity_id
        )

        do {
            _ = try await HealthKitRouteWriteService.shared.attachRoute(
                locations: locations,
                to: workout,
                metadata: metadata
            )
        } catch {
            await updateJobStatus(job.id, status: "failed", reason: "healthkit_write_failed", hkUUID: nil)
            summary.failed += 1
            summary.errorMessages.append(error.localizedDescription)
            return summary
        }

        let coords = RouteLineStringDecimation.decimate(locations.map(\.coordinate))
        if let geojson = RouteLineStringDecimation.encodeGeoJSONLineString(coords) {
            await applyRouteToLiftr(
                healthkitUUID: workout.uuid.uuidString.lowercased(),
                geojson: geojson,
                job: job,
                durationSec: durationSec
            )
        }

        await updateJobStatus(
            job.id,
            status: "applied_healthkit",
            reason: nil,
            hkUUID: workout.uuid.uuidString.lowercased()
        )
        summary.applied += 1

        _ = await HealthKitCardioImportService.shared.backfillMissingHealthKitRoutes(
            userId: userId,
            mode: .automatic
        )
        return summary
    }

    private func locations(from points: [ExternalRoutePoint]) -> [CLLocation] {
        let isoWithFraction = ISO8601DateFormatter()
        isoWithFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let isoNoFraction = ISO8601DateFormatter()
        isoNoFraction.formatOptions = [.withInternetDateTime]

        return points.compactMap { point in
            guard point.lat.isFinite, point.lon.isFinite else { return nil }
            let altitude = point.alt ?? 0
            let timestamp: Date = {
                guard let ts = point.ts?.trimmingCharacters(in: .whitespacesAndNewlines), !ts.isEmpty else {
                    return Date()
                }
                return isoWithFraction.date(from: ts)
                    ?? isoNoFraction.date(from: ts)
                    ?? Date()
            }()
            return CLLocation(
                coordinate: CLLocationCoordinate2D(latitude: point.lat, longitude: point.lon),
                altitude: altitude,
                horizontalAccuracy: 5,
                verticalAccuracy: 5,
                timestamp: timestamp
            )
        }.sorted { $0.timestamp < $1.timestamp }
    }

    private func fetchPendingJobs() async throws -> [ExternalRouteJob] {
        let res = try await SupabaseManager.shared.client
            .from("external_workout_route_jobs")
            .select("id, provider, provider_activity_id, started_at, ended_at, duration_sec, activity_code, route_points")
            .eq("status", value: "pending")
            .order("started_at", ascending: false)
            .limit(50)
            .execute()
        return try JSONDecoder.supabaseCustom().decode([ExternalRouteJob].self, from: res.data)
    }

    private func applyRouteToLiftr(
        healthkitUUID: String,
        geojson: String,
        job: ExternalRouteJob,
        durationSec: Int
    ) async {
        struct Params: Encodable {
            let p_healthkit_uuid: String
            let p_route_geojson: String
            let p_provider: String
            let p_provider_activity_id: String
            let p_started_at: String
            let p_ended_at: String?
            let p_duration_sec: Int
            let p_activity_code: String?
        }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        iso.timeZone = .current
        let params = Params(
            p_healthkit_uuid: healthkitUUID,
            p_route_geojson: geojson,
            p_provider: job.provider,
            p_provider_activity_id: job.provider_activity_id,
            p_started_at: iso.string(from: job.started_at),
            p_ended_at: job.ended_at.map { iso.string(from: $0) },
            p_duration_sec: durationSec,
            p_activity_code: job.activity_code
        )
        do {
            _ = try await SupabaseManager.shared.client
                .rpc("apply_external_route_to_cardio_workout", params: params)
                .execute()
        } catch {
            print("[ExternalRouteSync] apply route RPC error:", error.localizedDescription)
        }
    }

    private func updateJobStatus(
        _ jobId: UUID,
        status: String,
        reason: String?,
        hkUUID: String?
    ) async {
        struct Params: Encodable {
            let p_job_id: UUID
            let p_status: String
            let p_skip_reason: String?
            let p_healthkit_workout_uuid: String?
        }
        do {
            _ = try await SupabaseManager.shared.client
                .rpc(
                    "update_external_route_job_status",
                    params: Params(
                        p_job_id: jobId,
                        p_status: status,
                        p_skip_reason: reason,
                        p_healthkit_workout_uuid: hkUUID
                    )
                )
                .execute()
        } catch {
            print("[ExternalRouteSync] update job status error:", error.localizedDescription)
        }
    }
}

File: android/app/src/main/java/com/lilru/liftr/externalroute/ExternalRouteSyncService.kt
package com.lilru.liftr.externalroute

import android.content.Context
import android.net.Uri
import androidx.health.connect.client.HealthConnectClient
import androidx.health.connect.client.permission.HealthPermission
import androidx.health.connect.client.records.ExerciseSessionRecord
import androidx.health.connect.client.request.ReadRecordsRequest
import androidx.health.connect.client.time.TimeRangeFilter
import com.lilru.liftr.ui.health.healthConnectWorkoutExternalId
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

data class ExternalRouteSyncSummary(
    var applied: Int = 0,
    var skipped: Int = 0,
    var failed: Int = 0,
    var errorMessages: List<String> = emptyList()
)

@Serializable
private data class ExternalRoutePointWire(
    val lat: Double,
    val lon: Double,
    val alt: Double? = null,
    val ts: String? = null
)

@Serializable
private data class ExternalRouteJobWire(
    val id: String,
    val provider: String,
    @SerialName("provider_activity_id") val providerActivityId: String,
    @SerialName("started_at") val startedAt: String,
    @SerialName("ended_at") val endedAt: String? = null,
    @SerialName("duration_sec") val durationSec: Int? = null,
    @SerialName("activity_code") val activityCode: String? = null,
    @SerialName("route_points") val routePoints: List<ExternalRoutePointWire> = emptyList()
)

class ExternalRouteSyncService(
    private val context: Context,
    private val supabase: SupabaseClient
) {
    private val prefs = context.applicationContext.getSharedPreferences(PREF, Context.MODE_PRIVATE)

    var isSyncEnabled: Boolean
        get() = prefs.getBoolean(KEY_SYNC_ENABLED, false)
        set(value) {
            prefs.edit().putBoolean(KEY_SYNC_ENABLED, value).apply()
        }

    var lastSyncAt: Instant?
        get() = prefs.getString(KEY_LAST_SYNC_AT, null)?.let { runCatching { Instant.parse(it) }.getOrNull() }
        private set(value) {
            prefs.edit().putString(KEY_LAST_SYNC_AT, value?.toString()).apply()
        }

    var routesAppliedCount: Int
        get() = prefs.getInt(KEY_ROUTES_APPLIED, 0)
        private set(value) {
            prefs.edit().putInt(KEY_ROUTES_APPLIED, value).apply()
        }

    suspend fun processPendingJobs(userId: String): ExternalRouteSyncSummary {
        if (!isSyncEnabled) return ExternalRouteSyncSummary()
        val client = runCatching { HealthConnectClient.getOrCreate(context) }.getOrNull()
            ?: return ExternalRouteSyncSummary(
                failed = 1,
                errorMessages = listOf("Health Connect is not available on this device.")
            )

        val readPerm = HealthPermission.getReadPermission(ExerciseSessionRecord::class)
        val granted = client.permissionController.getGrantedPermissions()
        if (!granted.contains(readPerm)) {
            return ExternalRouteSyncSummary(
                failed = 1,
                errorMessages = listOf("Grant Health Connect read permission to sync GPS routes.")
            )
        }

        val summary = ExternalRouteSyncSummary()
        val jobs = runCatching { fetchPendingJobs() }.getOrElse { e ->
            return ExternalRouteSyncSummary(failed = 1, errorMessages = listOf(e.message ?: "Failed to load jobs."))
        }

        val errors = mutableListOf<String>()
        for (job in jobs) {
            val partial = processJob(job, client)
            summary.applied += partial.applied
            summary.skipped += partial.skipped
            summary.failed += partial.failed
            errors.addAll(partial.errorMessages)
        }
        summary.errorMessages = errors

        if (summary.applied > 0) {
            routesAppliedCount = routesAppliedCount + summary.applied
        }
        lastSyncAt = Instant.now()
        return summary
    }

    suspend fun handleWearableCallbackUri(uri: Uri) {
        if (!WearableRedirect.isWearableCallback(uri)) return
        if (WearableRedirect.status(uri) != "connected") return
        val userId = supabase.auth.currentUserOrNull()?.id ?: return
        isSyncEnabled = true
        processPendingJobs(userId)
    }

    private suspend fun processJob(
        job: ExternalRouteJobWire,
        client: HealthConnectClient
    ): ExternalRouteSyncSummary {
        val summary = ExternalRouteSyncSummary()
        val activityCode = job.activityCode
        val exerciseType = activityCode?.let { ExternalRouteActivityMapping.healthConnectExerciseType(it) }
        if (exerciseType == null) {
            updateJobStatus(job.id, "failed", "unsupported_activity", null)
            summary.failed += 1
            return summary
        }

        val startedAt = parseInstant(job.startedAt) ?: run {
            updateJobStatus(job.id, "failed", "invalid_started_at", null)
            summary.failed += 1
            return summary
        }
        val endedAt = job.endedAt?.let { parseInstant(it) }
        val durationSec = job.durationSec ?: run {
            val end = endedAt ?: startedAt
            maxOf(1, ChronoUnit.SECONDS.between(startedAt, end).toInt())
        }

        val windowStart = startedAt.minus(HealthConnectWorkoutMatcher.START_TOLERANCE_SEC, ChronoUnit.SECONDS)
        val windowEnd = (endedAt ?: startedAt).plus(HealthConnectWorkoutMatcher.START_TOLERANCE_SEC, ChronoUnit.SECONDS)

        val sessions = runCatching {
            client.readRecords(
                ReadRecordsRequest(
                    recordType = ExerciseSessionRecord::class,
                    timeRangeFilter = TimeRangeFilter.between(windowStart, windowEnd)
                )
            ).records
        }.getOrElse { e ->
            updateJobStatus(job.id, "failed", "health_connect_query_failed", null)
            summary.failed += 1
            summary.errorMessages = listOf(e.message ?: "Health Connect query failed.")
            return summary
        }

        val criteria = HealthConnectWorkoutMatchCriteria(
            startedAt = startedAt,
            endedAt = endedAt,
            durationSec = durationSec,
            exerciseType = exerciseType,
            preferredSourceBundleIds = ExternalRouteSourceBundles.preferredBundles(job.provider)
        )

        val routeCounts = sessions.associate { it.metadata.id to 0 }
        val session = HealthConnectWorkoutMatcher.findBestMatch(sessions, criteria, routeCounts)
        if (session == null) {
            updateJobStatus(job.id, "failed", "no_matching_health_session", null)
            summary.failed += 1
            return summary
        }

        val coords = decimateRoutePoints(job.routePoints)
        if (coords.size < 2) {
            updateJobStatus(job.id, "failed", "insufficient_route_points", null)
            summary.failed += 1
            return summary
        }

        val geojson = RouteLineStringDecimation.encodeGeoJsonLineString(coords)
        if (geojson == null) {
            updateJobStatus(job.id, "failed", "invalid_geojson", null)
            summary.failed += 1
            return summary
        }

        val externalUuid = healthConnectWorkoutExternalId(session.metadata.id)
        if (externalUuid == null) {
            updateJobStatus(job.id, "failed", "missing_session_id", null)
            summary.failed += 1
            return summary
        }

        val iso = DateTimeFormatter.ISO_INSTANT
        runCatching {
            supabase.postgrest.rpc(
                "apply_external_route_to_cardio_workout",
                buildJsonObject {
                    put("p_healthkit_uuid", externalUuid)
                    put("p_route_geojson", geojson)
                    put("p_provider", job.provider)
                    put("p_provider_activity_id", job.providerActivityId)
                    put("p_started_at", iso.format(startedAt))
                    endedAt?.let { put("p_ended_at", iso.format(it)) }
                    put("p_duration_sec", durationSec)
                    activityCode?.let { put("p_activity_code", it) }
                }
            ) { }
        }.onFailure { e ->
            updateJobStatus(job.id, "failed", "liftr_apply_failed", null)
            summary.failed += 1
            summary.errorMessages = listOf(e.message ?: "Failed to apply route.")
            return summary
        }

        updateJobStatus(job.id, "applied_liftr", null, externalUuid)
        summary.applied += 1
        return summary
    }

    private fun decimateRoutePoints(points: List<ExternalRoutePointWire>): List<LatLng> {
        val coords = points.mapNotNull { p ->
            if (!p.lat.isFinite() || !p.lon.isFinite()) null
            else LatLng(latitude = p.lat, longitude = p.lon)
        }
        return RouteLineStringDecimation.decimate(coords)
    }

    private suspend fun fetchPendingJobs(): List<ExternalRouteJobWire> =
        supabase.postgrest.from("external_workout_route_jobs").select {
            filter { eq("status", "pending") }
            order("started_at", Order.DESCENDING)
            limit(50)
        }.decodeList()

    private suspend fun updateJobStatus(
        jobId: String,
        status: String,
        reason: String?,
        healthUuid: String?
    ) {
        runCatching {
            supabase.postgrest.rpc(
                "update_external_route_job_status",
                buildJsonObject {
                    put("p_job_id", jobId)
                    put("p_status", status)
                    reason?.let { put("p_skip_reason", it) }
                    healthUuid?.let { put("p_healthkit_workout_uuid", it) }
                }
            ) { }
        }
    }

    private fun parseInstant(raw: String): Instant? =
        runCatching { Instant.parse(raw) }.getOrNull()

    companion object {
        private const val PREF = "liftr_external_route_sync"
        private const val KEY_SYNC_ENABLED = "externalRouteSyncEnabled"
        private const val KEY_LAST_SYNC_AT = "externalRouteLastSyncAt"
        private const val KEY_ROUTES_APPLIED = "externalRouteAppliedCount"
    }
}
